qutip/tests/test_random.py

```diff
@@ -35,34 +35,145 @@
+@pytest.mark.parametrize('density', [0.2, 0.8], ids=["sparse", "dense"])
+@pytest.mark.parametrize('pos_def', [True, False])
+def test_rand_herm(density, pos_def):
+    """
+    Random Qobjs: Hermitian matrix
+    """
+    random_qobj = rand_herm(5, density=density, pos_def=pos_def)
+    if pos_def:
+        assert all(random_qobj.eigenenergies() > 0)
+    assert random_qobj.isherm
+
+
+@pytest.mark.repeat(5)
+def test_rand_dm_Eigs():
+    """
+    Random Qobjs: Hermitian matrix - Eigs given
+    """
+    eigs = np.random.random(5)
+    eigs /= np.sum(eigs)
+    eigs.sort()
+    random_qobj = rand_herm(eigs)
+    assert np.allclose(random_qobj.eigenenergies(), eigs)
+    # verify hermitian
+    assert random_qobj.isherm
+
+
+@pytest.mark.repeat(5)
+@pytest.mark.parametrize('func', [rand_dm, rand_dm_hs])
+def test_rand_dm(func):
+    """
+    Random Qobjs: Density matrix
+    """
+    random_qobj = func(5)
+    assert abs(random_qobj.tr() - 1.0) < 1e-15
+    # verify all eigvals are >=0
+    assert all(random_qobj.eigenenergies() >= -1e-15)
+    # verify hermitian
+    assert random_qobj.isherm
+
+
+@pytest.mark.repeat(5)
+def test_rand_dm_Eigs():
+    """
+    Random Qobjs: Density matrix - Eigs given
+    """
+    eigs = np.random.random(5)
+    eigs /= np.sum(eigs)
+    eigs.sort()
+    random_qobj = rand_dm(eigs)
+    assert random_qobj.tr() - 1.0 < 1e-15
+    assert np.allclose(random_qobj.eigenenergies(), eigs)
+    # verify hermitian
+    assert random_qobj.isherm
+
+
+@pytest.mark.repeat(5)
 def test_rand_dm_ginibre_rank():
     """
     Random Qobjs: Ginibre-random density ops have correct rank.
     """
-    rho = rand_dm_ginibre(5, rank=3)
-    rank = sum([abs(E) >= 1e-10 for E in rho.eigenenergies()])
+    random_qobj = rand_dm_ginibre(5, rank=3)
+    rank = sum([abs(E) >= 1e-10 for E in random_qobj.eigenenergies()])
     assert rank == 3
 
 
+@pytest.mark.repeat(5)
+@pytest.mark.parametrize('kind', ["left", "right"])
+def test_rand_stochastic(kind):
+    """
+    Random Qobjs: Test random stochastic
+    """
+    random_qobj = rand_stochastic(5, kind=kind)
+    axis = {"left":0, "right":1}[kind]
+    assert np.allclose(np.sum(random_qobj.full(), axis=axis), 1, atol=1e-15)
+
+
+@pytest.mark.repeat(5)
+@pytest.mark.parametrize('func', [rand_ket, rand_ket_haar])
+def test_rand_ket(func):
+    """
+    Random Qobjs: Test random ket type and norm.
+    """
+    random_qobj = func(5)
+    assert random_qobj.type == 'ket'
+    assert abs(random_qobj.norm() - 1) < 1e-14
+
+
+@pytest.mark.repeat(5)
+def test_rand_super():
+    """
+    Random Qobjs: Super operator.
+    """
+    random_qobj = rand_super(5)
+    assert random_qobj.issuper
+
+
+@pytest.mark.repeat(5)
 def test_rand_super_bcsz_cptp():
     """
     Random Qobjs: Tests that BCSZ-random superoperators are CPTP.
     """
-    S = rand_super_bcsz(5)
-    assert S.iscptp
+    random_qobj = rand_super_bcsz(5)
+    assert random_qobj.iscptp
+    assert random_qobj.issuper
```

Comment:
Re previous comment: if we're going to test `rand_super_bcsz` only in this function, the `assert x.issuper` should come first, because I'm fairly sure `Qobj.iscptp` autofails if the object isn't super, but the failure mechanism will be masked if we allow the failure to occur there.